v0.28: Faceting settings (#1761)
meilisearch/documentation: reference/api/faceting.md

Faceting settings can also be updated directly through the [global settings route](/reference/api/settings.md#update-settings) along with the other settings.

To learn more about filtering and faceting, refer to our [dedicated guide](/learn/advanced/filtering_and_faceted_search.md).

::: warning
Updating the settings means overwriting the default settings of Meilisearch. You can reset to default values using the `DELETE` routes.
:::

## Get faceting settings

<RouteHighlighter method="GET" route="/indexes/{index_uid}/settings/faceting"/>

Get the faceting settings of an index. The index [`uid`](/learn/core_concepts/indexes.md#index-uid) is required.

### Example

<CodeSamples id="get_faceting_settings_1" />

#### Response: `200 OK`

```json
{
  "maxValuesPerFacet": 100
}
```

### Returned fields

#### `maxValuesPerFacet`

Maximum number of facet values returned for each facet.

## Update faceting settings

<RouteHighlighter method="PATCH" route="/indexes/{index_uid}/settings/faceting"/>

Partially update the faceting settings for an index. The index [`uid`](/learn/core_concepts/indexes.md#index-uid) is required.

### Body

#### `maxValuesPerFacet`

**Type:** integer

> **Default value:** `100`

Reviewer: Should we mention that 100 is the recommended value? From the [spec (lines 31-33)](https://github.com/meilisearch/specifications/blob/fb8b067925afda0ec51855df87a165a50222da5b/text/157-faceting-setting-api.md#311-maxvaluesperfacet)

>The value of 100 ensures good performance and prevents malicious users from scraping data from a Meilisearch instance.
>Increasing this value can degrade performance as well as expose the data of an instance to scrapping.

To learn more about filtering and faceting, refer to our [dedicated guide](/learn/advanced/filtering_and_faceted_search.md).

::: warning
Updating the settings means overwriting the default settings of Meilisearch. You can reset to default values using the `DELETE` routes.
:::

## Get faceting settings

<RouteHighlighter method="GET" route="/indexes/{index_uid}/settings/faceting"/>

Get the faceting settings of an index. The index [`uid`](/learn/core_concepts/indexes.md#index-uid) is required.

### Example

<CodeSamples id="get_faceting_settings_1" />

#### Response: `200 OK`

```json
{
  "maxValuesPerFacet": 100
}
```

### Returned fields

#### `maxValuesPerFacet`

Maximum number of facet values returned for each facet.

## Update faceting settings

<RouteHighlighter method="PATCH" route="/indexes/{index_uid}/settings/faceting"/>

Partially update the faceting settings for an index. The index [`uid`](/learn/core_concepts/indexes.md#index-uid) is required.

### Body

#### `maxValuesPerFacet`

**Type:** integer
**Default value:** `100`

> Configure the maximum number of facet values returned for each facet. Values are alphabetically sorted.

Reviewer 2: I am not sure that we should specify that the results are returned in this order, as it is not a specified behavior, that's just how the engine works right now and not how the specs want it.

Author: I think the spec might have been recently updated: https://github.com/meilisearch/specifications/pull/157/files#diff-6738a194cb97b82d2f0051b59a87cb03a343c5bad665ee11c2da0faf00a273c8R35

Unless I'm misunderstanding what we mean by lexicographical order on L#35?

Reviewer 2: We mean that if there were capital letters, for example, the capital one would be before the lowercase versions. It really applies to the raw values encoded as UTF-8. The real order is the order in terms of byte representation and not determined by human things like alphabetic or numeric order. For example `2` will be listed before `1000`.

Reviewer 3: Indeed, the actual behavior has been updated in the specification because I had internal questions about it.

I think it's not bad to explain it, it also allows people to know about it and to tell us if there are frustrations about it so we can improve it later. (my 2 cents)

Author: Thanks for the clarification. I was hoping we could have gotten away with alphabetical order since it's a bit more user-friendly, but, indeed, it sacrifices too much technical accuracy. 

I have accepted @gmourier's suggestion changing "alphabetical" to "ascending lexicographical" as I also think some users might appreciate knowing the expected behaviour.

Thanks to both of you for the assistance!

> Configure the maximum number of facet values returned for each facet. Values are alphabetically sorted.

Reviewer 3: ```suggestion
Configure the maximum number of facet values returned for each facet. Values are sorted in ascending lexicographical order.
```

For example, suppose a query's search results contain a total of three values for a `colors` facet: `blue`, `green`, and `red`. If you set `maxValuesPerFacet` to `2`, Meilisearch will only return `blue` and `green` in the response body's `facetDistribution` object.

::: note

> Settings `maxValuesPerFacet` to a high value might negatively impact performance.

Reviewer: ```suggestion
Setting `maxValuesPerFacet` to a high value might negatively impact performance.
```

:::

#### Example

<CodeSamples id="update_faceting_settings_1" />

#### Response: `202 Accepted`

```json
{
  "taskUid": 1,
  "indexUid": "books",
  "status": "enqueued",
  "type": "settingsUpdate",
  "enqueuedAt": "2022-04-14T20:56:44.991039Z"
}
```

> You can use the returned `uid` to get more details on [the status of the task](/reference/api/tasks.md#get-task).

Reviewer: ```suggestion
You can use the returned `taskUid` to get more details on [the status of the task](/reference/api/tasks.md#get-task).
```

## Reset faceting settings

Reset an index's faceting settings to their default value. The index [`uid`](/learn/core_concepts/indexes.md#index-uid) is required.

#### Example

<CodeSamples id="reset_faceting_settings_1" />

#### Response: `200 OK`

```json
{
  "taskUid": 1,
  "indexUid": "books",
  "status": "enqueued",
  "type": "settingsUpdate",
  "enqueuedAt": "2022-04-14T20:53:32.863107Z"
}
```

> You can use the returned `uid` to get more details on [the status of the task](/reference/api/tasks.md#get-task).

Reviewer: ```suggestion
You can use the returned `taskUid` to get more details on [the status of the task](/reference/api/tasks.md#get-task).
```
